mixnet/node/src/lib.rs

```diff
@@ -77,205 +76,279 @@ impl MixnetNode {
-            Body::new_sphinx(packet),
-            NymNodeRoutingAddress::try_from(next_node_addr)?,
-        )
-        .await
+        self.forward(packet, NymNodeRoutingAddress::try_from(next_node_addr)?)
+            .await
     }
 
     async fn forward_payload_to_destination(
-        pool: &ConnectionPool,
-        max_retries: usize,
-        retry_delay: Duration,
-        payload: Payload,
+        &self,
+        payload: Body,
         destination_addr: DestinationAddressBytes,
     ) -> Result<()> {
         tracing::debug!("Forwarding final payload to destination mixnode");
 
-        Self::forward(
-            pool,
-            max_retries,
-            retry_delay,
-            Body::new_final_payload(payload),
+        self.forward(
+            payload,
             NymNodeRoutingAddress::try_from_bytes(&destination_addr.as_bytes())?,
         )
         .await
     }
 
-    async fn forward(
-        pool: &ConnectionPool,
-        max_retries: usize,
-        retry_delay: Duration,
-        body: Body,
-        to: NymNodeRoutingAddress,
-    ) -> Result<()> {
+    async fn forward(&self, msg: Body, to: NymNodeRoutingAddress) -> Result<()> {
         let addr = SocketAddr::from(to);
-        let arc_socket = pool.get_or_init(&addr).await?;
-
-        if let Err(e) = {
-            let mut socket = arc_socket.lock().await;
-            body.write(&mut *socket).await
-        } {
-            tracing::error!("Failed to forward packet to {addr} with error: {e}. Retrying...");
-            return mixnet_protocol::retry_backoff(
-                addr,
-                max_retries,
-                retry_delay,
-                body,
-                arc_socket,
-            )
-            .await
-            .map_err(Into::into);
-        }
+
+        self.message_tx.send(TargetedMessage::new(addr, msg))?;
         Ok(())
     }
 }
+
+struct MessageHandler {
+    // TODO: remove this allow when we implement the retry logic
+    #[allow(dead_code)]
+    config: MixnetNodeConfig,
+    message_rx: mpsc::UnboundedReceiver<TargetedMessage>,
+    message_tx: mpsc::UnboundedSender<TargetedMessage>,
+    connections: HashMap<SocketAddr, TcpStream>,
+}
+
+impl MessageHandler {
+    pub fn new(
+        message_tx: mpsc::UnboundedSender<TargetedMessage>,
+        message_rx: mpsc::UnboundedReceiver<TargetedMessage>,
+        config: MixnetNodeConfig,
+    ) -> Self {
+        Self {
+            message_tx,
+            message_rx,
+            connections: HashMap::with_capacity(config.connection_pool_size),
+            config,
+        }
+    }
+
+    pub async fn run(mut self) {
+        loop {
+            tokio::select! {
+                msg = self.message_rx.recv() => {
+                    if let Some(msg) = msg {
+                        match self.send(msg).await {
+                            Ok(Some(msg)) => {
+                                let _ = self.message_tx.send(msg);
+                            }
```

Comment:
This is an example of why the signature of the function it's hard to read, also, a comment explaining this re-queues the message would be helpful, as it's not explicit where the other half of the channel is